### PR TITLE
Inference pipeline

### PR DESCRIPTION
**Pull request by saagar-parikh**
Add functions to implement the following sequentially: get all ids of a field, get all features of the ids, run inference for all pre-trained models on all ids of the field

### CODE REVIEW

File: tools/inference.py [hunk 1]
```diff
@@ -108,65 +171,284 @@ def make_model(**kwargs):
...
+                + str(round(te - ts, 4))
+                + " s"
+            )
+        preds_df = features[["_id", model_class + '_dnn']].round(2)
+        preds_df.reset_index(inplace=True, drop=True)
+    else:
+        # xgboost inferencing
+        train_config = path_model[-7]
+        feature_names = config["inference"]["xgb"][train_config]
+
```
Comment (bfhealy) [inline tools/inference.py:374]: Should clean_dataset_xgb be called here to prepare the features?

File: tools/scope_download_classification.py [hunk 2]
```diff
@@ -174,4 +174,4 @@ def download_classification(file: str, gloria, group_ids: list, token: str):
     args = parser.parse_args()
 
     # download object classifications in the file
-    download_classification(args.file, gloria, args.group_ids, args.token)
+    download_classification(args.file, gloria, args.group_ids, args.token)
```
Comment (bfhealy) [inline tools/scope_download_classification.py:177]: Should there be a newline at the end of this file?

**Review by bfhealy — CHANGES_REQUESTED** [review]
Thank you for submitting these very useful upgrades! I have left some remarks, mostly consisting of comments/questions about how the code runs. There are also a few minor change requests.

File: get_all_preds.sh [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+echo -n "pnp ... "  && python tools/inference.py --path-model=models/dr5-1/pnp-20210918_041303.h5   --model-class=pnp --field=$1 --whole-field && echo "done"
+echo -n "puls ... " && python tools/inference.py --path-model=models/dr5-1/puls-20210920_012555.h5  --model-class=puls --field=$1 --whole-field && echo "done"
+echo -n "rrlyr ... " && python tools/inference.py --path-model=models/dr5-1/rrlyr-20210920_041205.h5 --model-class=rrlyr --field=$1 --whole-field && echo "done"
+echo -n "rscvn ... " && python tools/inference.py --path-model=models/dr5-1/rscvn-20210920_045003.h5 --model-class=rscvn --field=$1 --whole-field && echo "done"
+echo -n "srv ... "  && python tools/inference.py --path-model=models/dr5-1/srv-20210920_053008.h5   --model-class=srv --field=$1 --whole-field && echo "done"
+echo -n "vnv ... "  && python tools/inference.py --path-model=models/dr5-1/vnv-20210915_220725.h5   --model-class=vnv --field=$1 --whole-field && echo "done"
+echo -n "yso ... "  && python tools/inference.py --path-model=models/dr5-1/yso-20210920_184534.h5   --model-class=yso --field=$1 --whole-field && echo "done"
```
[the remark this answers is not in the document]
Comment (saagar-parikh) [reply get_all_preds.sh:23]: I have added those classes in the script for which the trained models were available to me, so all of them are present in a dir on schoty. Also, I have not uploaded any new labels to Gloria; the labels in ZTF_source_classifications_DR5 were already present.

File: tools/get_features.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+        query = {
+            "query_type": "find",
+            "query": {
+                "catalog": features_catalog,
+                "filter": {"_id": {"$in": source_ids}},
+            },
+            "kwargs": {"limit": limit, "skip": skip},
+        }
+        response = kowalski.query(query=query)
+        source_data = response.get("data")
```
[the remark this answers is not in the document]
Comment (saagar-parikh) [reply tools/get_features.py:73]: I have not experimented much with setting the limit, but for limit=1000 for the ZTF_source_features_DR5 catalog, I was able to avoid the timeout issue.

File: get_all_preds.sh [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+echo -n "puls ... " && python tools/inference.py --path-model=models/d15.puls.c.model --model-class=puls --xgb-model --field=$1 --whole-field && echo "done"
+echo -n "rrlyr ... " && python tools/inference.py --path-model=models/d15.rrlyr.f.model --model-class=rrlyr --xgb-model --field=$1 --whole-field && echo "done"
+echo -n "rscvn ... " && python tools/inference.py --path-model=models/d15.rscvn.f.model --model-class=rscvn --xgb-model --field=$1 --whole-field && echo "done"
+echo -n "srv ... "  && python tools/inference.py --path-model=models/d15.srv.f.model --model-class=srv --xgb-model --field=$1 --whole-field && echo "done"
+echo -n "vnv ... "  && python tools/inference.py --path-model=models/d15.vnv.c.model --model-class=vnv --xgb-model --field=$1 --whole-field && echo "done"
+echo -n "wuma ... "  && python tools/inference.py --path-model=models/d15.wuma.f.model --model-class=wuma --xgb-model --field=$1 --whole-field && echo "done"
+echo -n "yso ... "  && python tools/inference.py --path-model=models/d15.yso.f.model --model-class=yso --xgb-model --field=$1 --whole-field && echo "done"
```
[the remark this answers is not in the document]
Comment (saagar-parikh) [reply get_all_preds.sh:46]: Yes, these classes could be included in the script if needed (their trained models are also available). I did not include them because there were no available corresponding dnn models. Should I include these in the script now?

File: tools/get_quad_ids.py [hunk 6]
```diff
@@ -1,8 +1,219 @@
...
+        How many of the selected rows to return. Default is 10000
+    output_dir : str
+        Relative directory path to save output files to
+
+    Returns
+    -------
+    Directory containing hdf5 files for each quad in the specified range
+
+    USAGE: get_all_ids(get_field_ids, 'ZTF_sources_20210401',field=301,ccd_range=2,quad_range=4,\
+        minobs=5,limit=2000)
```
[the remark this answers is not in the document]
Comment (saagar-parikh) [reply tools/get_quad_ids.py:159]: Yes, I will make the changes.

File: tools/get_quad_ids.py [hunk 7]
```diff
@@ -56,60 +267,150 @@ def get_field_ids(catalog, field=301, ccd=4, quad=3, minobs=20, skip=0, limit=10
     r = gloria.query(q)
```
[the remark this answers is not in the document]
Comment (saagar-parikh) [reply tools/get_quad_ids.py:267]: Since this query only extracts the ids, the default limit is set to 10000 - no timeout issues were faced for this limit. I will update the docstring.

File: tools/inference.py (hunk 1 above)
Comment (saagar-parikh) [reply tools/inference.py:374]: I did not keep it then because all the features that contained missing values were not inputs to the xgb models. However, I now realize it is better to keep a safety clean_dataset function to prepare the features in case of future updates to either the model inputs or the dataset. Should I add it now?

File: tools/scope_download_classification.py (hunk 2 above)
Comment (saagar-parikh) [reply tools/scope_download_classification.py:177]: Yes. I had overlooked the "no newline" error earlier.

File: tools/get_quad_ids.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (saagar-parikh) [reply tools/get_quad_ids.py:159]: Yes, I will change the docstring accordingly.

**Comment by saagar-parikh** [comment]
Thank you for the useful feedback and suggestions in the code. I will update the code and docstrings for functions wherever required soon.

File: get_all_preds.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (bfhealy) [reply get_all_preds.sh:46]: Thanks, you can go ahead and include them. I think it's best to include all the available models, especially since we can comment out those lines if we don't want to use them in any given run.

File: tools/inference.py (hunk 1 above)
Comment (bfhealy) [reply tools/inference.py:374]: Sounds good, probably best to add it for the future like you mentioned.

File: tools/get_quad_ids.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (bfhealy) [reply tools/get_quad_ids.py:159]: Thanks!

File: get_all_preds.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (bfhealy) [reply get_all_preds.sh:23]: Ok, thanks for the info.

**Comment by bfhealy** [comment]
Thank you Saagar! It will be nice to have these useful updates added to scope.

File: get_all_preds.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (AshishMahabal) [reply get_all_preds.sh:46]: I agree. @saagar-parikh downstream we were requiring models that were implementted in xgboost and dnn, but that logic can stay separate for those kinds of specific requirements.

**Comment by bfhealy** [comment]
Hi @saagar-parikh, just wanted to check how the changes are coming along.

**Comment by bfhealy** [comment]
Those changes look good @saagar-parikh! Just one more step before merging - could you please address the linting errors below?

tools/get_features.py:1:1: F401 'tarfile.DEFAULT_FORMAT' imported but unused
tools/get_features.py:7:1: F401 'warnings' imported but unused
tools/get_features.py:8:1: F401 'typing.Union' imported but unused
tools/get_features.py:10:1: F401 'json' imported but unused
tools/get_features.py:95:17: E712 comparison to False should be 'if cond is False:' or 'if not cond:'
tools/get_features.py:178:17: E712 comparison to False should be 'if cond is False:' or 'if not cond:'

**Comment by saagar-parikh** [comment]
@bfhealy I will check the flake8 tests and make the changes. But I'm not able to understand the Kowalski not found error. Can you please guide?

**Comment by bfhealy** [comment]
@saagar-parikh The Kowalski error should be ok, some recent PRs passed the checks and still displayed that error. I think it failed this time because black found some changes to make. You can run those checks locally before committing:

`pre-commit install`
`pre-commit run --files <file1> <file2> <file3>`

Then the formatting revisions will be made locally, after which you can commit them one more time.

**Comment by bfhealy** [comment]
(You'll have to pip install pre-commit first if you don't have it already)

**Comment by bfhealy** [comment]
Hi @saagar-parikh, any luck with pre-commit?

**Comment by saagar-parikh** [comment]
Hi @bfhealy, I think it's just the Kowalski error that's left now!

**Comment by bfhealy** [comment]
Thanks very much @saagar-parikh! I'll go ahead and merge.
